src/quick-lint-js/fe/parse-statement.cpp

```diff
@@ -417,6 +418,20 @@ bool Parser::parse_and_visit_statement(Parse_Visitor_Base &v,
       this->skip();
       this->check_body_after_label();
       goto parse_statement;
+    } else if (this->peek().type == Token_Type::kw_function) {
+      this->parse_and_visit_function_declaration(
+          v, Function_Declaration_Options{
+                 .attributes = Function_Attributes::normal,
+                 .begin = this->peek().begin,
+                 .require_name = Name_Requirement::required_for_statement,
+                 .async_keyword = std::nullopt,
+                 .declare_keyword = std::nullopt,
+             });
+      if (this->peek().type != Token_Type::left_paren) {
+        this->diag_reporter_->report(
+            Diag_Unexpected_Await_On_Function_Declaration{
+                .await_keyword = await_token.span()});
+      }
```

Comment:
Must fix: Your special handling of `function` is wrong in a few ways:

* `await function foo() {}();` is parsed incorrectly. You have a test for it, but `test_parse_and_visit_statement` doesn't assert that the whole input is parsed. (It probably should... I should fix that.) This code is parsing `await function foo() {}` as a statement, leaving `();` as a second statement (which is wrong!).
* `await function () {}` falsely reports E0061 ("missing name in function statement").

Also, adding code here in parse_and_visit_expression doesn't catch mistakes such as the following:
```javascript
foo(await function() {
});
```

`parse_await_expression` is probably a better place to put your logic.